### PR TITLE
[ast] AST update

### PR DESCRIPTION
**Pull request by arnonsha**
Signed-off-by: Arnon Sharlin <arnon.sharlin@opentitan.org>

### CODE REVIEW

**Comment by arnonsha** [comment]
2nd AST delivery. Changes :

    New files (22 vs. 20:
    - ana_pkg.sv // new nettype, awire
    - ast_pkg.sv // All delay/other parameters of AST moved here.
    - rglts_pdm_3p3v.sv // Replaces main_rglt.sv & pdm.sv

    Add `ifndef VERILATOR to pass Verilator.

    Add `ifndef SYNTHESIS to pass FPGA & LINT.

    General code updates in all files (logic and CodingStyle issues)

    Change interface definition of ADC analog input vector to non-paced format.
    From: input wire [AdcChannels-1:0] adc_ai,
    To: input awire/wire adc_a0_ai and
    input awire/wire adc_a1_ai

    Enable Real Number Modeling (RNM) for ADC using of new ‘nettype’, awire, from ana_pkg.sv.

    Update other AST interface signals (dft and otp).

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/data/ast.hjson:7]: just curious, why is a different comment format used here? both work, but we typically just stick to `//` for uniformity...

File: hw/top_earlgrey/ip/ast/rtl/aon_osc.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/aon_osc.sv:32]: can this be `always_comb`?

File: hw/top_earlgrey/ip/ast/rtl/ast_pkg.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/ast_pkg.sv:17]: nit: these should all be indented by 2 spaces according to our style.

File: hw/top_earlgrey/ip/ast/rtl/gen_pok.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/gen_pok.sv:38]: `always_comb`?

File: hw/top_earlgrey/ip/ast/rtl/io_osc.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/io_osc.sv:34]: `always_comb`?

File: hw/top_earlgrey/ip/ast/rtl/lc_ctrl_pkg.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/lc_ctrl_pkg.sv:6]: So this file already exists on master `hw/ip/lc_ctrl/rtl/` and hence does not have to be checked in again. 

You can add a dependency on the life cycle controller package to your AST core file. I'll highlight that with a separate comment.

File: hw/top_earlgrey/ip/ast/rtl/rglts_pdm_3p3v.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/rglts_pdm_3p3v.sv:26]: this signal should also have an `_o` suffix.

File: hw/top_earlgrey/ip/ast/rtl/rng_osc.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/rng_osc.sv:36]: `always_comb`?

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:41]: `always_comb`?

File: hw/top_earlgrey/ip/ast/ast.core
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/ast.core:10]: you can add `- lowrisc:ip:lc_ctrl_pkg` here in order to get the definitions for the life-cycle-related types.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: I believe the `pwrmgr` still depends on that file, and that is also why all CI checks fail.

@tjaychen is this still needed or can we remove `ast_wrapper_pkg.sv` from the `pwrmgr` dependencies (or replace it with something else)?

**Review by msfschaffner — COMMENTED** [review]
Thanks @arnonsha! This looks mostly good, but I have a couple of comments related to style and SV constructs. 

I.e. I noticed that `always @(*)` is used in several places. We are trying to avoid this legacy construct and hence I was wondering whether we could replace all those instances with `always_comb`?

Also, there is unneeded white-space at the end of several lines. It would be good to configure the editor to automatically trim that away, since I believe our style linter will complain about it, hence preventing us from merging into master. 

The current CI failure is due to a core-file dependency issue, though - see below.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: o boy i completely forgot to review this.

@msfschaffner is right, if we want to remove ast_wrapper then all the thing packages that were defined by the ast for integration convenience need to be filled in elsewhere.   

@arnonsha did you intend to remove these?

File: hw/top_earlgrey/ip/ast/ast.core
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/ast.core:10]: There's no lc_ctrl_pkg in v1-bronze. This is abypass I agreed on with @tjaychen. WIll be fixed in SLVR.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: IIRC @tjaychen at some point you have said you would remove the wrapper in Silver.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:7]: I think we took an example from one of the existing files, but we will fix
This hybrid comment format also appears in the following files:

./hw/dv/data/xcelium/xcelium.hjson
./hw/ip/aes/data/aes.hjson
./hw/ip/alert_handler/data/alert_handler.hjson
./hw/ip/clkmgr/data/clkmgr.hjson
./hw/ip/nmi_gen/data/nmi_gen.hjson
./hw/ip/padctrl/data/padctrl.hjson
./hw/ip/pinmux/data/pinmux.hjson
./hw/ip/rstmgr/data/rstmgr.hjson
./hw/ip/rv_plic/data/rv_plic.hjson
./hw/ip/sensor_ctrl/data/sensor_ctrl.hjson
./hw/top_earlgrey/ip/alert_handler/data/autogen/alert_handler.hjson
./hw/top_earlgrey/ip/clkmgr/data/autogen/clkmgr.hjson
./hw/top_earlgrey/ip/padctrl/data/autogen/padctrl.hjson
./hw/top_earlgrey/ip/pinmux/data/autogen/pinmux.hjson
./hw/top_earlgrey/ip/rv_plic/data/autogen/rv_plic.hjson
./hw/vendor/lowrisc_ibex/dv/uvm/data/xcelium/xcelium.hjson

File: hw/top_earlgrey/ip/ast/rtl/aon_osc.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/aon_osc.sv:32]: the always_comb is only good for synthesizable logic. In our case this is behavioral.
If you'll put always_comb VCS will fail compilation. This is good since the tools will let you know that what's written in the always cannot be synthesized to a combinational logic (same for always_ff by the way).

So this none-synthesizable logic uses always @(*)

File: hw/top_earlgrey/ip/ast/rtl/ast_pkg.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_pkg.sv:17]: Will fix

File: hw/top_earlgrey/ip/ast/rtl/io_osc.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/io_osc.sv:34]: see my comment on aon_osc.v

File: hw/top_earlgrey/ip/ast/rtl/rglts_pdm_3p3v.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/rglts_pdm_3p3v.sv:26]: will fix

File: hw/top_earlgrey/ip/ast/rtl/rng_osc.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/rng_osc.sv:36]: see my comment on aon_osc.v

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: yes we can remove it.  But the contents of ast_wrapper_pkg.sv do need to be reproduced elsewhere though, because the rest of the system relies on it for integration purposes.  

Would that be okay with you?

Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: OK from our side @tjaychen 
If you plan to do it before SLVR, please also make the same change to the bronze branch so we have same core file in both master and bronze.

Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: sounds good.
for now, since this is going to master branch, could you restore the ast_wrapper related functions? Otherwise this PR will always fail top level CIs.

Then as a separate PR we can work on removing the wrapper and using ast directly.

Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: Sorry, I am not sure exactly what should be done:
Is it juest removing the "      - lowrisc:ip:pwrmgr_pkg" dependency from the ast_wrapper_pkg ?

Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: I think you just have to restore the `ast_wrapper_pkg.sv` file to make this work (this PR deletes it).

File: hw/top_earlgrey/ip/ast/ast.core
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/ast.core:10]: I agree for `v1-bronze`, but this PR is merged into `master`, right? the package is available on master...

Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/ast.core:10]: @tjaychen just to clarify: should this PR go into `master` or `v1-bronze`?

File: hw/top_earlgrey/ip/ast/rtl/aon_osc.sv
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/aon_osc.sv:32]: ok makes sense - thanks for the clarification!

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:7]: that would be great, thanks!

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: no not yet, because there are a few other modules I think that also depend on it.
So when we remove ast_wrapper_pkg, we need to go through and clean-up all those references. Which is why I think it should be a separate PR for now.

What do you think?

File: hw/top_earlgrey/ip/ast/ast.core
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/ast.core:10]: This should go into master.
Once approved I will do another PR for bronze.
I think it is best to have the same ast configuration in both master and bronze-v1. So for now I'd like to keep this file in ast even though it appears elsewhere.
Once lc is added to bronze-v1 I'll remove it.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv:1]: This is strange @msfschaffner , I really don't understand how this pkg was deleted (also the ast pkg). I can see them in my forked repo.
Hope it will be restored with my next commit.

Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: OK, the ast_wrapper pkg is back with the new commit. Hope this solves the CI issue

**Comment by tjaychen** [comment]
arnon do you also mind filing an opentitan issue to remove the ast_wrapper
constructs?
otherwise i might just forget.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: ok, thanks, that brings us one step further. now it looks like the ast core file is still referencing `main_rglt.sv` which is deleted in this PR also. hence that should be removed from the core file for consistency.

I think in order to speed this process up a little bit it would be good if you could run style linting on your side and weed out the compilation and style issues in order to unblock CI.
this should be really lightweight so set up since it uses the opensource tool Verible, which is easy to install (you can just download a precompiled binary). 

I.e., follow the steps [here](https://docs.opentitan.org/doc/ug/install_instructions/#verible), and run the following two regressions locally:
```
$ util/dvsim/dvsim.py -t veriblelint hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson
$ util/dvsim/dvsim.py -t veriblelint hw/top_earlgrey/lint/top_earlgrey_dv_lint_cfgs.hjson
```
these two regressions are enabled in CI, which basically means that if they don't pass we cannot merge into master.

Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: Sorry @msfschaffner  about the core. WIll fix.

Regarding verible lint. I have 2 syntax errors on legit system verilog code, which I couldn't find a way to waive off:

1.  hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:55:6: syntax error, rejected "#" (https://github.com/google/verible).
Line 56 is:
`wire #(USB_VAL_RDLY, USB_VAL_FDLY) ref_val = usb_ref_val_i;
Here we define rise delay and  fall delay for this wire (remember this is behavioral code)

2. hw/top_earlgrey/ip/ast//rtl/ana_pkg.sv:17:3: syntax error, rejected "nettype" (https://github.com/google/verible).
We use a legit SV nettype syntax (used for real number modeling)

Please advise on how to bypass

**Comment by arnonsha** [comment]
@tjaychen 
I opened issue #4050

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: @arnonsha, can you push the latest state onto this PR such that I can take a look? It could be that those constructs are not supported by Verible yet (in which case I will waive them and file a bug on Verible).

Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: thanks @msfschaffner. updated.
ast.core file is now fixed, and ast pass vcs compilation

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:56]: @arnonsha ok if we rewrite this in this way?

**Comment by msfschaffner** [comment]
Ok thanks @arnonsha, I've pulled this PR down and performed the following cleanup steps:
- remove all trailing whitespace
- remove the duplicate `lc_ctrl_pkg` since that broke our VCS regressions
- rewrite wire assignment to circumvent syntax error in Verilator
This should make CI pass now.

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:56]: Yes, that is OK

File: hw/top_earlgrey/ip/ast/ast.core
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/ast.core]: FYI, I had to remove this since it otherwise collides with the already existing `lc_ctrl_pkg`. For instance, this broke our VCS regression. I'd rather suggest we add the "real" `lc_ctrl_pkg` to bronze instead.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (fangism) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: I'm not sure how you have the linter setup, but you can't waive syntax errors (valid or invalid), so you'll need to make sure you have another way to bypass to tolerate the occasional tool bug.  (Thanks for filing https://github.com/google/verible/issues/581)

File: hw/top_earlgrey/ip/ast/ast.core
Comment (arnonsha) [reply hw/top_earlgrey/ip/ast/ast.core]: Yes, adding the pkg the bronze was also my initial suggestion.
So once this PR is merged to master, and the lc_ctrl_pkg is added to v1-bronze we can push this PR to bronze as well.

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper.sv
Comment (msfschaffner) [inline hw/top_earlgrey/ip/ast/rtl/ast_wrapper.sv:66]: @tjaychen PTAL

File: hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv
Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: I think we currently let Verible back off in CI if there is a Syntax error, since that would otherwise block the PR...

Comment (msfschaffner) [reply hw/top_earlgrey/ip/ast/rtl/ast_wrapper_pkg.sv]: Sorry I should add that initially I thought these where lint errors that could be waived. But you are right that the waiver mechanism does of course not work if the error is due to syntax.

**Comment by msfschaffner** [comment]
Ok this seems to pass CI now. Merging in!

**Review by msfschaffner — APPROVED** [review]

**Pull request merged**
